code/game/machinery/engi_points_vendor.dm

```diff
@@ -0,0 +1,696 @@
+			if(glow.light_color == "se")
+				glow.light_color = "#ffff00"
+				glow.light_range = glow.range * 0.9
+				glow.update_light()
+				continue
+			if(glow.light_color == "#ffff00")
+				glow.light_color = "ne"
+				glow.light_range = 0
+				glow.update_light()
+				continue
+			if(glow.light_color == "ne")
+				glow.light_color = "red"
+				glow.light_power = glow.light_power * 0.68
+				glow.light_range = glow.range * 0.85
+				glow.update_light()
+				continue
+		if(prob(2))
+			INVOKE_ASYNC(src, .proc/hierofunk)
+		sleep(selection.song_beat)
+
+
+/obj/machinery/disco/proc/dance(var/mob/living/carbon/M) //Show your moves
+	switch(rand(0,9))
+		if(0 to 1)
+			set waitfor = 0
+			for(var/i = 1, i < 8, i++)
+				M.SpinAnimation(7,1)
+				M.setDir(pick(cardinal))
+				sleep(10)
+		if(2 to 3)
+			set waitfor = 0
+			for(var/i in 1 to 6)
+				if (!M)
+					return
+				M.SpinAnimation(7,1)
+				M.setDir(pick(cardinal))
+				for (var/x in 1 to 12)
+					sleep(1)
+					if (!M)
+						return
+					if (i<5)
+						M.pixel_y += 1
+					if (i>4)
+						M.pixel_y -= 2
+					M.setDir(turn(M.dir, 90))
+					switch (M.dir)
+						if (NORTH)
+							M.pixel_y += 3
+						if (SOUTH)
+							M.pixel_y -= 3
+						if (EAST)
+							M.pixel_x -= 3
+						if (WEST)
+							M.pixel_x += 3
+				sleep(12)
+			M.pixel_x = 0
+			M.pixel_y = 0
+		if(4 to 6)
+			var/speed = rand(1,3)
+			set waitfor = 0
+			var/time = 30
+			while(time)
+				sleep(speed)
+				for(var/i in 1 to speed)
+					M.setDir(pick(cardinal))
+					M.lay_down(TRUE)
+				 time--
+
+		if(7 to 9)
+			M.setDir(get_dir(M, src))
+			spawn (0)
+				if (M)
+					animate(M, transform = matrix(180, MATRIX_ROTATE), time = 1, loop = 0)
+				sleep (70)
+				if (M)
+					animate(M, transform = null, time = 1, loop = 0)
+			for (var/i = 0, i < 60, i++)
+				if (!M)
+					return
+				if (i<31)
+					M.pixel_y += 1
+				if (i>30)
+					M.pixel_y -= 1
+				M.setDir(turn(M.dir, 90))
+				switch (M.dir)
+					if (NORTH)
+						M.pixel_y += 3
+					if (SOUTH)
+						M.pixel_y -= 3
+					if (EAST)
+						M.pixel_x -= 3
+					if (WEST)
+						M.pixel_x += 3
+				sleep (1)
+			M.pixel_x = 0
+			M.pixel_y = 0
+
+/obj/machinery/disco/proc/dance_over()
+	for(var/obj/item/device/flashlight/spotlight/SL in spotlights)
+		qdel(SL)
+	spotlights.Cut()
+	for(var/obj/effect/overlay/sparkles/SP in sparkles)
+		qdel(SP)
+	sparkles.Cut()
+	rangers.Cut()
+	for(var/mob/living/L in listeners)
+		if(!L || !L.client)
+			continue
+		L.client.stop_client_sounds()
+	listeners.Cut()
+
+
+/obj/machinery/disco/process()
+	if(charge<35)
+		charge += 1
+	if(world.time < stop && active)
+		rangers = list()
+		for(var/mob/living/M in range(9,src))
+			rangers += M
+			if(!(M in listeners))
+				M << selection.song_path
+				listeners += M
+			if(prob(4+(allowed(M)*4)))
+				dance(M)
+		for(var/mob/living/L in listeners)
+			if(!(L in rangers))
+				listeners -= L
+				if(!L || !L.client)
+					continue
+				L.client.stop_client_sounds()
```

Comment:
I screwed around for many hours trying to get a better way but in testing it worked better than I'd hoped. Yea having it cut out instead of fading or something isn't ideal but it wasn't that jarring either. The play_sound or whatever system also wasn't usable because even without any pitch variation the "atmospheric/distance" effects make any music sound horrible.